File: skills/ow-hero-sync-guard/scripts/player_constants_reachability.py
#!/usr/bin/env python3
"""Dependency-aware reachability scan for player_constants defines.

Default mode is report-only. Deletion requires explicit flags:
- scoped delete: --prefix <HERO_> --apply
- global delete: --global-cleanup --apply
"""

from __future__ import annotations

import argparse
import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Set

DEFINE_RE = re.compile(r"^#!define\s+([A-Z][A-Z0-9_]*)\s+(.*)$")
TOKEN_RE = re.compile(r"\b[A-Z][A-Z0-9_]*\b")
COMMENT_RE = re.compile(r"^\s*#")


@dataclass
class ScanResult:
    defined: int
    roots: int
    reachable: int
    unused_total: int
    candidate_unused: List[str]


def parse_defines(constants_file: Path) -> tuple[List[str], Dict[str, str], List[str]]:
    lines = constants_file.read_text(encoding="utf-8").splitlines()
    order: List[str] = []
    expr: Dict[str, str] = {}

    for line in lines:
        match = DEFINE_RE.match(line)
        if not match:
            continue
        name, value = match.group(1), match.group(2)
        order.append(name)
        expr[name] = value

    return order, expr, lines


def build_dependencies(expr: Dict[str, str], defs: Set[str]) -> Dict[str, Set[str]]:
    deps: Dict[str, Set[str]] = {}
    for name, value in expr.items():
        tokens = set(TOKEN_RE.findall(value))
        deps[name] = {tok for tok in tokens if tok in defs and tok != name}
    return deps


def collect_external_tokens(src_root: Path, constants_file: Path) -> Set[str]:
    tokens: Set[str] = set()
    for path in src_root.rglob("*.opy"):
        if path.resolve() == constants_file.resolve():
            continue
        for line in path.read_text(encoding="utf-8").splitlines():
            if COMMENT_RE.match(line):
                continue
            tokens.update(TOKEN_RE.findall(line))
    return tokens


def reachable_constants(
    defs: Set[str], deps: Dict[str, Set[str]], external_tokens: Set[str]
) -> Set[str]:
    roots = defs & external_tokens
    keep: Set[str] = set()
    stack = list(roots)

    while stack:
        cur = stack.pop()
        if cur in keep:
            continue
        keep.add(cur)
        stack.extend(deps.get(cur, set()) - keep)

    return keep


def filter_candidates(unused: List[str], prefixes: List[str], global_cleanup: bool) -> List[str]:
    if global_cleanup:
        return unused
    if not prefixes:
        return []
    return [name for name in unused if any(name.startswith(p) for p in prefixes)]


def apply_delete(lines: List[str], delete_names: Set[str], constants_file: Path) -> None:
    out: List[str] = []
    for line in lines:
        match = DEFINE_RE.match(line)
        if match and match.group(1) in delete_names:
            continue
        out.append(line)

    constants_file.write_text("\n".join(out) + "\n", encoding="utf-8")


def run(args: argparse.Namespace) -> int:
    constants_file = Path(args.constants_file)
    src_root = Path(args.src_root)

    if not constants_file.exists():
        raise FileNotFoundError(f"constants file not found: {constants_file}")
    if not src_root.exists():
        raise FileNotFoundError(f"src root not found: {src_root}")

    if args.apply and not args.global_cleanup and not args.prefix:
        raise ValueError("refuse to apply without scope: use --prefix <HERO_> or --global-cleanup")

    order, expr, lines = parse_defines(constants_file)
    defs = set(order)
    deps = build_dependencies(expr, defs)
    external_tokens = collect_external_tokens(src_root, constants_file)
    keep = reachable_constants(defs, deps, external_tokens)

    unused = [name for name in order if name not in keep]
    candidates = filter_candidates(unused, args.prefix, args.global_cleanup)

    result = ScanResult(
        defined=len(order),
        roots=len(defs & external_tokens),
        reachable=len(keep),
        unused_total=len(unused),
        candidate_unused=candidates,
    )

    payload = {
        "defined": result.defined,
        "external_roots": result.roots,
        "reachable": result.reachable,
        "unused_total": result.unused_total,
        "candidate_unused_count": len(result.candidate_unused),
        "candidate_unused": result.candidate_unused,
        "constants_file": str(constants_file),
        "scope": {
            "prefix": args.prefix,
            "global_cleanup": args.global_cleanup,
            "apply": args.apply,
        },
    }

    if args.json:
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print(f"defined={result.defined}")
        print(f"external_roots={result.roots}")
        print(f"reachable={result.reachable}")
        print(f"unused_total={result.unused_total}")
        print(f"candidate_unused_count={len(result.candidate_unused)}")
        for name in result.candidate_unused:
            print(name)

    if args.report_file:
        Path(args.report_file).write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    if args.apply and result.candidate_unused:
        apply_delete(lines, set(result.candidate_unused), constants_file)
        print(f"applied_delete_count={len(result.candidate_unused)}")
    elif args.apply:
        print("applied_delete_count=0")

    return 0


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Reachability scan for player_constants defines")
    parser.add_argument(
        "--constants-file",
        default="src/constants/player_constants.opy",
        help="path to constants file",
    )
    parser.add_argument(
        "--src-root",
        default="src",
        help="source root for external reference scan",
    )
    parser.add_argument(
        "--prefix",
        action="append",
        default=[],
        help="limit candidate deletion to constants starting with prefix (repeatable)",
    )
    parser.add_argument(
        "--global-cleanup",
        action="store_true",
        help="allow candidate set to include all unused constants",
    )
    parser.add_argument(
        "--apply",
        action="store_true",
        help="apply deletion to constants file",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="print report in JSON format",
    )
    parser.add_argument(
        "--report-file",
        default="",
        help="optional JSON report output path",
    )
    return parser


if __name__ == "__main__":
    parser = build_parser()
    ns = parser.parse_args()
    raise SystemExit(run(ns))

File: tools/player-constants-reachability.ts
#!/usr/bin/env -S node --import tsx

import fs from "node:fs/promises";
import path from "node:path";
import { repoRoot, resolveRepo } from "./lib/runtime.js";

const DEFINE_RE = /^#!define\s+([A-Z][A-Z0-9_]*)\s+(.*)$/;
const TOKEN_RE = /\b[A-Z][A-Z0-9_]*\b/g;
const COMMENT_RE = /^\s*#/;

type Args = {
  constantsFile: string;
  srcRoot: string;
  prefixes: string[];
  globalCleanup: boolean;
  apply: boolean;
  json: boolean;
  reportFile: string | null;
};

function parseArgs(argv: string[]): Args {
  const args: Args = {
    constantsFile: "src/constants/player_constants.opy",
    srcRoot: "src",
    prefixes: [],
    globalCleanup: false,
    apply: false,
    json: false,
    reportFile: null,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index]!;
    switch (arg) {
      case "--constants-file":
        args.constantsFile = argv[++index] ?? "";
        break;
      case "--src-root":
        args.srcRoot = argv[++index] ?? "";
        break;
      case "--prefix":
        args.prefixes.push(argv[++index] ?? "");
        break;
      case "--global-cleanup":
        args.globalCleanup = true;
        break;
      case "--apply":
        args.apply = true;
        break;
      case "--json":
        args.json = true;
        break;
      case "--report-file":
        args.reportFile = argv[++index] ?? "";
        break;
      case "-h":
      case "--help":
        printHelp();
        process.exit(0);
      default:
        throw new Error(`Unknown argument: ${arg}`);
    }
  }

  if (args.apply && !args.globalCleanup && args.prefixes.length === 0) {
    throw new Error("refuse to apply without scope: use --prefix <HERO_> or --global-cleanup");
  }

  return args;
}

function printHelp(): void {
  console.log("Usage: tools/player-constants-reachability.ts [options]");
  console.log("  --constants-file PATH");
  console.log("  --src-root PATH");
  console.log("  --prefix HERO_");
  console.log("  --global-cleanup");
  console.log("  --apply");
  console.log("  --json");
  console.log("  --report-file PATH");
}

async function parseDefines(constantsFile: string): Promise<{
  order: string[];
  expr: Map<string, string>;
  lines: string[];
}> {
  const lines = (await fs.readFile(constantsFile, "utf8")).split(/\r?\n/);
  const order: string[] = [];
  const expr = new Map<string, string>();

  for (const line of lines) {
    const match = line.match(DEFINE_RE);
    if (!match) {
      continue;
    }
    order.push(match[1]!);
    expr.set(match[1]!, match[2]!);
  }

  return { order, expr, lines };
}

function buildDependencies(expr: Map<string, string>, defs: Set<string>): Map<string, Set<string>> {
  const deps = new Map<string, Set<string>>();
  for (const [name, value] of expr.entries()) {
    const next = new Set<string>();
    for (const token of value.match(TOKEN_RE) ?? []) {
      if (defs.has(token) && token !== name) {
        next.add(token);
      }
    }
    deps.set(name, next);
  }
  return deps;
}

async function collectExternalTokens(srcRoot: string, constantsFile: string): Promise<Set<string>> {
  const tokens = new Set<string>();
  async function walk(current: string): Promise<void> {
    const entries = await fs.readdir(current, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(current, entry.name);
      if (entry.isDirectory()) {
        await walk(fullPath);
        continue;
      }
      if (!entry.name.endsWith(".opy") || path.resolve(fullPath) === path.resolve(constantsFile)) {
        continue;
      }
      const lines = (await fs.readFile(fullPath, "utf8")).split(/\r?\n/);
      for (const line of lines) {
        if (COMMENT_RE.test(line)) {
          continue;
        }
        for (const token of line.match(TOKEN_RE) ?? []) {
          tokens.add(token);
        }
      }
    }
  }
  await walk(srcRoot);
  return tokens;
}

function reachableConstants(
  defs: Set<string>,
  deps: Map<string, Set<string>>,
  externalTokens: Set<string>,
): Set<string> {
  const roots = [...defs].filter((name) => externalTokens.has(name));
  const keep = new Set<string>();
  const stack = [...roots];

  while (stack.length > 0) {
    const current = stack.pop()!;
    if (keep.has(current)) {
      continue;
    }
    keep.add(current);
    for (const dep of deps.get(current) ?? []) {
      if (!keep.has(dep)) {
        stack.push(dep);
      }
    }
  }

  return keep;
}

function filterCandidates(unused: string[], prefixes: string[], globalCleanup: boolean): string[] {
  if (globalCleanup) {
    return unused;
  }
  if (prefixes.length === 0) {
    return [];
  }
  return unused.filter((name) => prefixes.some((prefix) => name.startsWith(prefix)));
}

async function applyDelete(lines: string[], deleteNames: Set<string>, constantsFile: string): Promise<void> {
  const kept = lines.filter((line) => {
    const match = line.match(DEFINE_RE);
    return !match || !deleteNames.has(match[1]!);
  });
  await fs.writeFile(constantsFile, `${kept.join("\n")}\n`, "utf8");
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  const constantsFile = resolveRepo(args.constantsFile);
  const srcRoot = resolveRepo(args.srcRoot);

  const { order, expr, lines } = await parseDefines(constantsFile);
  const defs = new Set(order);
  const deps = buildDependencies(expr, defs);
  const externalTokens = await collectExternalTokens(srcRoot, constantsFile);
  const keep = reachableConstants(defs, deps, externalTokens);

  const unused = order.filter((name) => !keep.has(name));
  const candidates = filterCandidates(unused, args.prefixes, args.globalCleanup);

  const payload = {
    defined: order.length,
    external_roots: [...defs].filter((name) => externalTokens.has(name)).length,
    reachable: keep.size,
    unused_total: unused.length,
    candidate_unused_count: candidates.length,
    candidate_unused: candidates,
    constants_file: path.relative(repoRoot, constantsFile),
    scope: {
      prefix: args.prefixes,
      global_cleanup: args.globalCleanup,
      apply: args.apply,
    },
  };

  if (args.json) {
    console.log(JSON.stringify(payload, null, 2));
  } else {
    console.log(`defined=${payload.defined}`);
    console.log(`external_roots=${payload.external_roots}`);
    console.log(`reachable=${payload.reachable}`);
    console.log(`unused_total=${payload.unused_total}`);
    console.log(`candidate_unused_count=${payload.candidate_unused_count}`);
    for (const name of candidates) {
      console.log(name);
    }
  }

  if (args.reportFile) {
    await fs.writeFile(resolveRepo(args.reportFile), `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  }

  if (args.apply && candidates.length > 0) {
    await applyDelete(lines, new Set(candidates), constantsFile);
    console.log(`applied_delete_count=${candidates.length}`);
  } else if (args.apply) {
    console.log("applied_delete_count=0");
  }
}

main().catch((error: Error) => {
  console.error(error.message);
  process.exit(1);
});
